File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/src/DBSarrayFn.c
//  
//
//  Created by James Ramsay on 2018-3-6.
//
//

#include <stdio.h>

void DBSarrayFn(int    *nXbasisw, int    *nWbasisw, int    *nUbasisj, int *nAbasisj,
                int    *nrep,     double *Avecj,    double *Ucoefj,
                double *BAtenswj, double *DBSarray)
{
    
    int nXw = *nXbasisw;
    int nWw = *nWbasisw;
    int nUj = *nUbasisj;
    int nAj = *nAbasisj;
    int nRp = *nrep;
    
    int nDBSarray = nXw*nRp*nWw;
    
    /* assign 0 to all positions */
    
    for (int p = 0; p < nDBSarray; p++)
    {
        DBSarray[p]=0;
    }
    
    /* loop through all columns of all four input matrices */
    
    for (int r = 0; r < nRp; r++)
    {
        for (int i = 0; i < nXw; i++)
        {
            for (int j = 0; j < nWw; j++)
            {
                for (int k = 0; k < nUj; k++)
                {
                        for (int l = 0; l < nAj; l++)
                    {
                        int ijkl = i*nAj*nUj*nWw + j*nAj*nUj + k*nAj + l;
                        
                        /*  increment value at index by the product */
                        
                        DBSarray[i+r*nXw+j*nRp*nXw] +=
                               Avecj[l] * Ucoefj[k+r*nUj] * BAtenswj[ijkl];
                        
                    }
                }
            }
        }
    }
}

File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/R/DBSarrayFn.R
DBSarrayFn <- function(nXbasisw, nWbasisw, 
                      nUbasisj, nAbasisj, 
                      nrep, Avecj, Ucoefj, BAtens) {
  BAtens <- as.vector(BAtens)
  DSvecC <- rep(0,nXbasisw*nrep*nWbasisw)
  inprodDBList <- .C("DBSarrayFn", as.integer(nXbasisw), 
                     as.integer(nWbasisw), 
                     as.integer(nUbasisj), 
                     as.integer(nAbasisj), 
                     as.integer(nrep),
                     as.double(Avecj),
                     as.double(Ucoefj),
                     as.double(BAtens), 
                     as.double(DSvecC))
  DBSarrayC <- inprodDBList[[9]]
  return(DBSarrayC)
}
